src/Analysis/Ast/Impl/Extensions/PythonPropertyExtensions.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Linq;
+using Microsoft.Python.Analysis.Analyzer;
+using Microsoft.Python.Analysis.Diagnostics;
+using Microsoft.Python.Analysis.Types;
+using Microsoft.Python.Core;
+using Microsoft.Python.Parsing;
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Analysis.Extensions {
+    public static class PythonPropertyExtensions {
+        /// <summary>
+        /// Returns true if the decorator combiantions for the property are valid
+        /// </summary>
+        public static bool HasValidDecorators(this IPythonPropertyType p, IExpressionEvaluator eval) {
+            bool valid = true;
+            // If property is abstract, allow all decorators because it will be overridden
+            if(p.IsAbstract) {
+                return valid;
+            }
+
+            foreach (var dec in (p.FunctionDefinition?.Decorators?.Decorators).MaybeEnumerate().OfType<NameExpression>()) {
```

Comment:
Since we already walk decorators and set flags in `ProcessDecorators`, do we need to do it again? Or can we just check odd combinations of the flags set? I.e. `IsStatic && IsClassMethod` then report error?

Reply:
I loop over decorators so I can highlight their locations. If it was fine to just generically highlight the function definition instead, then I can use the flag method, but if I want to highlight the decorators themselves I need their locations.